File: solution/17. Letter Combinations of a Phone Number/17_02.cpp
class Solution {
public:
    vector<string> letterCombinations(string digits) {
        if (digits.empty()) return {};
        
        const vector<string> m = {
            "0", "1", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"
        };
        
        vector<string> ans;
        ans.push_back("");
        for (char& d : digits) {
            vector<string> tmp;
            for (char c : m[d - '0']) {
                for (string& pre : ans) {
                    tmp.push_back(pre + c);
                }
            }
            swap(ans, tmp);
        }
        
        return ans;
    }
};

File: solution/17. Letter Combinations of a Phone Number/17_02.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        if len(digits) == 0:
            return []
        
        m = {
            '2': 'abc',
            '3': 'def',
            '4': 'ghi',
            '5': 'jkl',
            '6': 'mno',
            '7': 'pqrs',
            '8': 'tuv',
            '9': 'wxyz'
        }
        res = []
        res.append("")

        for d in digits:
            tmp = []
            for prefix in res:
                for c in m[d]:
                    tmp.append(prefix + c)
            res = tmp
        return res
